File: 0236.Lowest-Common-Ancestor-of-a-Binary-Tree/step2_bottom_up.cpp

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
  public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
        if (root == nullptr) {
            return nullptr;
        }
        if (root == p || root == q) {
            return root;
        }
        TreeNode *left_found = lowestCommonAncestor(root->left, p, q);
        TreeNode *right_found = lowestCommonAncestor(root->right, p, q);
        if (left_found == nullptr) {
            return right_found;
        }
        if (right_found == nullptr) {
            return left_found;
        }
        return root;
    }
};

File: 0236.Lowest-Common-Ancestor-of-a-Binary-Tree/step2_bottom_up.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode | None, p: TreeNode, q: TreeNode
    ) -> TreeNode | None:
        if root is None:
            return None
        if root == p or root == q:
            return root

        left_found = self.lowestCommonAncestor(root.left, p, q)
        right_found = self.lowestCommonAncestor(root.right, p, q)

        if left_found is None:
            return right_found
        if right_found is None:
            return left_found
        return root
